### PR TITLE
Rework show status

### PR DESCRIPTION
**Pull request by fbuihuu**

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -4380,12 +4358,34 @@ bool manager_is_confirm_spawn_disabled(Manager *m) {
         return access("/run/systemd/confirm_spawn_disabled", F_OK) >= 0;
 }
 
+static bool manager_should_show_status(Manager *m, StatusType type) {
```
Comment (poettering) [inline src/core/manager.c:4415]: should → shall? 

@keszybz is our resident Engish language expert, what do you say?

File: src/core/manager.c [hunk 2]
```diff
@@ -2760,11 +2762,11 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
                 switch (sfsi.ssi_signo - SIGRTMIN) {
 
                 case 20:
-                        manager_set_show_status(m, SHOW_STATUS_YES, "signal");
+                        manager_set_show_status_overridden(m, SHOW_STATUS_YES, "signal");
```
Comment (poettering) [inline src/core/manager.c]: maybe call this manager_override_show_status()

File: src/core/manager.c [hunk 3]
```diff
@@ -4260,30 +4262,78 @@ void manager_recheck_journal(Manager *m) {
...
+                return m->show_status_overridden;
+
+        return m->show_status;
+}
+
+bool manager_get_show_status_on(Manager *m) {
+        assert(m);
+
+        return show_status_on(manager_get_show_status(m));
+}
```
Comment (poettering) [inline src/core/manager.c:4285]: is it worth having a helper for this one? sounds easier to just write this out literally

File: src/core/dbus-manager.c [hunk 4]
```diff
@@ -2487,6 +2537,7 @@ const sd_bus_vtable bus_manager_vtable[] = {
         SD_BUS_PROPERTY("Environment", "as", property_get_environment, 0, 0),
         SD_BUS_PROPERTY("ConfirmSpawn", "b", bus_property_get_bool, offsetof(Manager, confirm_spawn), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("ShowStatus", "b", property_get_show_status, 0, 0),
+        SD_BUS_WRITABLE_PROPERTY("WritableShowStatus", "s", property_get_show_status_string, property_set_show_status_string, 0, 0),
```
Comment (poettering) [inline src/core/dbus-manager.c]: i don't like this I must say. I think it would be nicer to add a method then. SetShowStatus() which optionally takes the empty string

File: src/core/dbus-manager.c [hunk 5]
```diff
@@ -242,6 +242,56 @@ static int property_get_show_status(
...
+                void *userdata,
+                sd_bus_error *error) {
+
+        Manager *m = userdata;
+
+        assert(m);
+        assert(bus);
+        assert(reply);
+
+        return sd_bus_message_append_basic(reply, 's', manager_get_show_status_on(m) ? "yes" : "no");
```
Comment (poettering) [inline src/core/dbus-manager.c]: `yes_no(…)`

**Comment by fbuihuu** [comment]
@poettering new version force pushed, PTAL.

File: src/core/manager.c [hunk 6]
```diff
@@ -3568,14 +3570,14 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
                         else
                                 m->honor_device_enumeration = b;
 
-                } else if ((val = startswith(l, "show-status="))) {
+                } else if ((val = startswith(l, "show-status-overridden="))) {
```
Comment (poettering) [inline src/core/manager.c:3573]: hmm, why did you rename the field? I mean, of course it has different semantics, but shouldn't we just leave the old name to do something somewhat reasonable during updates that to reexec? i.e. is there a reason behind renaming this serialization field besides "feels cleaner"?

File: src/core/manager.c [hunk 7]
```diff
@@ -4264,30 +4266,78 @@ void manager_recheck_journal(Manager *m) {
...
                   strna(show_status_to_string(mode)),
                   reason);
 
-        m->show_status = mode;
-
-        if (enabled)
+        if (show_status_on(mode))
                 (void) touch("/run/systemd/show-status");
         else
                 (void) unlink("/run/systemd/show-status");
```
Comment (poettering) [inline src/core/manager.c]: maybe move these four lines into a helper function of its own, and invoke it from both manager_set_show_status_overridden() and manager_set_show_status(), after all both functions muck with this flag file, and should really do it the same way

File: src/core/manager.c (hunk 6 above)
Comment (fbuihuu) [reply src/core/manager.c:3573]: Because  "show-status"  always reflects the value set in the configuration file. Therefore this value is and should always be reloaded from the config file.

Serializing this field was actually a bug because, as explained in the commit message, it locked the initial value read from the config file making impossible any later customization.

File: src/core/manager.c (hunk 7 above)
Comment (fbuihuu) [reply src/core/manager.c]: ok

File: src/core/manager.c (hunk 6 above)
Comment (poettering) [reply src/core/manager.c:3573]: I am not arguing which field should be serialized. I am argue under which name to do that. i.e. previously the value read from the config and the "overriden" value wasn't distuingished, now it is. When you runtime update your systemd from an old version to  anew version that includes your patch then this means we'd suddenly forget the overriden state, since we would not process the old `show-status` value at all, but that's where both the config and the overriden thing ended up in...

Hence I am proposing: read the show-status= serialization field into the show_status_overriden variable (instead of the show_status variable).

**Comment by poettering** [comment]
looks good, still wondering about the name of the serialization field. i.e. I'd use the old serializeation field name for the new variable, even if this means there is a discrepancy between the in-memory variable naming and the serialization field naming, see my comments above.

Or is there a good reason not do do it like that that I missed?

File: src/core/manager.c (hunk 6 above)
Comment (fbuihuu) [reply src/core/manager.c:3573]: What you propose would still be boggus.

There's no way to fix that properly because as you wrote the "overridden" state is lost after serializing.

So when deserializing we have only 2 options: either we assume that show-status value was overridden or it was not.

You suggest to assume that the value was overridden. That would still be problematic because if show-status value was actually not overridden and systemd was reexecuted into a new version then the old value would become suddenly the overridden value which would mean that any attempt to customize its value through the config file would not work. IOW the existing bug would not be fixed by updating systemd.

With the current patch, which assumes that the old value was not overridden, the previous use case would be fixed. However if show-status was actually overridden (via the signal API) after updating systemd the old value will still be restored but it will loose its "overridden" status. It means that the value could later be overridden by simply reloading the config.

Assuming that changing show-status via the signal API is pretty uncommon, especially since this API is pretty awkward to use, I would say that the first use case is more important to fix. Also the code is less confusing (or feel "cleaner" if you prefer).

**Comment by fbuihuu** [comment]
@poettering did my last comment make sense or do you still insist to read the show-status= serialization field ?
